src/ucp/core/ucp_ep.c

```diff
@@ -632,24 +634,23 @@ void ucp_ep_config_init(ucp_worker_h worker, ucp_ep_config_t *config)
 
                 if (context->config.ext.zcopy_thresh == UCS_CONFIG_MEMUNITS_AUTO) {
                     /* auto */
-                    zcopy_thresh = md_attr->reg_cost.overhead / (
-                                            (1.0 / context->config.ext.bcopy_bw) -
-                                            (1.0 / iface_attr->bandwidth) -
-                                            md_attr->reg_cost.growth);
-                    if (zcopy_thresh < 0) {
-                        config->am.zcopy_thresh      = SIZE_MAX;
-                        config->am.sync_zcopy_thresh = -1;
-                    } else {
-                        config->am.zcopy_thresh      = zcopy_thresh;
-                        config->am.sync_zcopy_thresh = zcopy_thresh;
+                    config->am.zcopy_thresh[0]      = UCS_CONFIG_MEMUNITS_AUTO;
+                    config->am.sync_zcopy_thresh[0] = UCS_CONFIG_MEMUNITS_AUTO;
+                    for (it = 1; it <= UCP_MAX_IOV; ++it) {
+                        zcopy_thresh = ucp_ep_config_get_zcopy_auto_thresh(
+                                           it, &md_attr->reg_cost,
+                                           context->config.ext.bcopy_bw,
+                                           iface_attr->bandwidth);
+                        config->am.sync_zcopy_thresh[it] = zcopy_thresh;
+                        config->am.zcopy_thresh[it]      = ucs_max(zcopy_thresh,
+                                                                   iface_attr->cap.am.min_zcopy);
                     }
                 } else {
-                    config->am.zcopy_thresh      = context->config.ext.zcopy_thresh;
-                    config->am.sync_zcopy_thresh = context->config.ext.zcopy_thresh;
+                    config->am.sync_zcopy_thresh[1] = context->config.ext.zcopy_thresh;
```

Comment:
The `sync_zcopy_thresh[1]` will be used in only one case - then `sync_zcopy_thresh[0]` is not AUTO. It will be used for all `iovcnt` (including bigger than `UCP_MAX_IOV`).
I see you proposed a bit different algorithm to handle these thresholds.